core/ir/aarch64/codec.c

```diff
@@ -9561,6 +9561,106 @@ encode_opnds_tbz(byte *pc, instr_t *instr, uint enc, decode_info_t *di)
     return ENCFAIL;
 }
 
+static inline uint
+decode_load_store_category(uint enc)
+{
+    uint category = DR_INSTR_CATEGORY_OTHER;
+    /* Calculation of category is based on C4.1 'A64 instruction set encoding'
+     * of ARM V8 Architecture reference manual
+     *  https://developer.arm.com/documentation/ddi0487/
+     *  The encoding is:
+     *
+     *  31    28      26  24  23 22 21                      11 10                    0
+     * | x x x x | x | x x x | x x | x x x x x x | x x x x | x x | x x x x x x x x x x |
+     * -----------   ----  -----   ---------------         -------
+     *     op0        op1   op2          op3                 op4
+     *                        ------
+     *                         opc
+     */
+    uint op0 = BITS(enc, 31, 28);
+    uint opc = BITS(enc, 23, 22);
+    if ((op0 & 0x3) == 0x3) { /* xx11 */
+        if (BITS(enc, 10, 10) == 1 && BITS(enc, 21, 21) == 1)
+            category = DR_INSTR_CATEGORY_LOAD;
+        else if (opc == 0 || (opc == 0x2 && BITS(enc, 26, 26) == 1))
+            category = DR_INSTR_CATEGORY_STORE;
+        else
+            category = DR_INSTR_CATEGORY_LOAD;
+    } else if ((op0 & 0x3) == 0 || (op0 & 0x3) == 0x2) { /* xx00, xx10 */
+        category =
+            (BITS(enc, 22, 22) == 0) ? DR_INSTR_CATEGORY_STORE : DR_INSTR_CATEGORY_LOAD;
+        if ((op0 & 0xc) == 0 && BITS(enc, 26, 26) == 1)
+            category |= DR_INSTR_CATEGORY_SIMD;
+    } else { /* xx01 */
+        if (BITS(enc, 24, 24) == 0)
+            category = DR_INSTR_CATEGORY_LOAD;
+        else if (BITS(enc, 21, 21) == 0)
+            category = (opc == 0) ? DR_INSTR_CATEGORY_STORE : DR_INSTR_CATEGORY_LOAD;
+        else if ((opc == 0x1 || opc == 0x3) && BITS(enc, 11, 10) == 0)
+            category = DR_INSTR_CATEGORY_LOAD;
+        else
+            category = DR_INSTR_CATEGORY_STORE;
+    }
+    return category;
+}
+
+static inline bool
+decode_category(uint enc, instr_t *instr)
+{
+    int category = DR_INSTR_CATEGORY_OTHER;
+    /* Calculation of category is based on C4.1 'A64 instruction set encoding'
+     * of ARM V8 Architecture reference manual
+     *  The encoding is:
+     *
+     *   31  30 29 28    25 24                                             0
+     * | x | x  x |x x x x | x x x x x x x x x x x x x x x x x x x x x x x x |
+     *             --------
+     *               op1
+     */
+
+    uint op1 = BITS(enc, 28, 25);
+    if ((BITS(enc, 31, 31) == 1 && op1 == 0) || op1 == 0x2) /* SME || SVE */
+        category = DR_INSTR_CATEGORY_SIMD;
+    else if (BITS(enc, 31, 31) == 0 && op1 == 0) /* op1 is 0 and 31 bit is 0 */
+        category = DR_INSTR_CATEGORY_UNCATEGORIZED;
+    else {
+        /*                       op1 - xxxx
```

Comment:
Might be worth adding a `DR_INSTR_CATEGORY_SVE`  with `op1` = `0010`. The codec supports all SVE instructions.

Reply:
Hi @kuhanov, are you handling `op1` = `0001` and `op1` = `x111` as being in the same category of SIMD vector instructions for the purposes of this PR?

Bear in mind that the Advanced SIMD (NEON) instructions are for _fixed width_ 128 bit vectors. The op1 = 0010 encoding class covers _scalable vector widths_ from 128 to 2048 bits depending on the hardware implementation. SVE is a separate encoding class from SIMD even though they both handle vectors. There are 337 SVE opcodes, not counting the latest SVE2 instructions.

Reply:
I would agree with @kuhanov on the cross-arch general _SIMD category grouping Neon and SVE for this PR.  I think the intended use cases for these categories are for more general purposes not tied to any one architecture.  Distinguishing individual ISA-specific characteristics does have its own uses cases but I think that would be more along the lines of FEATURE_ bits and would be a separate thing from these categories.